// luabundler/lib/commands/bundle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path_1 = require("path");
const command_1 = require("@oclif/command");
const luabundle_1 = require("luabundle");
const util_1 = require("../util");
const luaVersions = ['5.1', '5.2', '5.3', 'LuaJIT'];
class BundleCommand extends command_1.Command {
    async run() {
        const { args, flags } = this.parse(BundleCommand);
        const options = {
            luaVersion: flags.lua,
            isolate: flags.isolate,
            expressionHandler: (module, expression) => {
                const start = expression.loc.start;
                console.warn(`WARNING: Non-literal require found in '${module.name}' at ${start.line}:${start.column}`);
            },
        };
        if (flags.path.length > 0) {
            options.paths = flags.path;
        }
        let content;
        if (!args.file || args.file === '-') {
            content = luabundle_1.bundleString(await util_1.readStdin(), options);
        }
        else {
            content = luabundle_1.bundle(args.file, options);
        }
        if (flags.output) {
            const resolvedPath = path_1.resolve(flags.output);
            const resolvedDir = path_1.dirname(resolvedPath);
            if (!fs_1.existsSync(resolvedDir)) {
                fs_1.mkdirSync(resolvedDir, { recursive: true });
            }
            fs_1.writeFileSync(flags.output, content);
        }
        else {
            console.log(content);
        }
    }
}
exports.default = BundleCommand;
BundleCommand.description = 'Bundles a Lua file and its dependencies into a single file';
BundleCommand.flags = {
    help: command_1.flags.help({ char: 'h' }),
    isolate: command_1.flags.boolean({ char: 'i', description: 'Disables require() fallback at runtime for modules not found in the bundle.', default: false }),
    lua: command_1.flags.enum({ char: 'l', description: 'Lua syntax version', options: luaVersions, default: '5.3' }),
    output: command_1.flags.string({ char: 'o', description: 'Bundle output path' }),
    path: command_1.flags.string({ char: 'p', description: 'Add a require path pattern', multiple: true, required: true }),
    version: command_1.flags.version({ char: 'v' }),
};
BundleCommand.args = [
    {
        name: 'file',
        required: false,
        description: 'Lua entry-point module file path. May be omitted to read from stdin instead.',
    },
];
