// luabundler/lib/commands/unbundle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path_1 = require("path");
const command_1 = require("@oclif/command");
const luabundle_1 = require("luabundle");
const util_1 = require("../util");
class UnbundleCommand extends command_1.Command {
    async run() {
        const { args, flags } = this.parse(UnbundleCommand);
        const { extension, modules, output } = flags;
        const options = {
            rootOnly: !modules,
        };
        let unbundled;
        if (!args.file || args.file === '-') {
            unbundled = luabundle_1.unbundleString(await util_1.readStdin(), options);
        }
        else {
            unbundled = luabundle_1.unbundle(args.file, options);
        }
        if (modules) {
            const modulesDir = path_1.resolve(modules);
            if (!fs_1.existsSync(modulesDir)) {
                fs_1.mkdirSync(modulesDir, { recursive: true });
            }
            for (const module of Object.values(unbundled.modules)) {
                if (module.name !== unbundled.metadata.rootModuleName) {
                    const modulePath = path_1.resolve(modulesDir, module.name).replace(/\./g, path_1.sep) + (extension ? '.' + extension : '');
                    const moduleDir = path_1.dirname(modulePath);
                    if (!fs_1.existsSync(moduleDir)) {
                        fs_1.mkdirSync(moduleDir, { recursive: true });
                    }
                    fs_1.writeFileSync(modulePath, module.content);
                }
            }
        }
        const rootModule = unbundled.modules[unbundled.metadata.rootModuleName];
        if (output) {
            const outputPath = path_1.resolve(output);
            const outputDir = path_1.dirname(outputPath);
            if (!fs_1.existsSync(outputDir)) {
                fs_1.mkdirSync(outputDir, { recursive: true });
            }
            fs_1.writeFileSync(outputPath, rootModule.content);
        }
        else {
            console.log(rootModule.content);
        }
    }
}
exports.default = UnbundleCommand;
UnbundleCommand.description = 'Unbundles a Lua bundle into individual files';
UnbundleCommand.flags = {
    help: command_1.flags.help({ char: 'h' }),
    modules: command_1.flags.string({ char: 'm', description: 'Output directory path where modules should be written.' }),
    extension: command_1.flags.string({ char: 'e', description: 'File extension of output modules. Only relevant when --modules (-m) is specified.', default: 'lua' }),
    output: command_1.flags.string({ char: 'o', description: 'Output path of the root (entry point) module.' }),
    version: command_1.flags.version({ char: 'v' }),
};
UnbundleCommand.args = [
    {
        name: 'file',
        required: false,
        description: 'Lua bundle file path. May be omitted to read from stdin instead.',
    },
];
